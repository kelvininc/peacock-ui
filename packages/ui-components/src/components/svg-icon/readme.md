# *<kv-svg-icon>*

## Styling Svg Icon Element

The `kv-svg-icon` element consists only of an icon that is displayed on the view. To customize this, style using a combination of CSS and any of the [CSS custom properties](#css-custom-properties):

```css
kv-svg-icon {
  /* Applies to the icon color */
  --icon-color: #103d73;

  /* Set a different icon height value */
  --icon-heigh: 16px;
  
  /* Set a different icon width value */
  --icon-heigh: 16px;

  /* Set a different icon rotation */
  --icon-rotation: 90deg;
}
```

Alternatively, depending on the [browser support](https://caniuse.com/#feat=mdn-css_selectors_part) needed, CSS shadow parts can be used to style the select:

```css
/* Set the width and height of icon to the full container */
kv-svg-icon::part(icon) {
  width: 100%;
  height: 100%;
}

/* Set the icon color */
kv-svg-icon::part(icon) {
  fill: kv-color(success);
}
```
Notice that by using `::part`, any CSS property on the element can be targeted.  

Additionally, we already have some predefined classes that can be applied to scale and rotate the icon. 

**Sizes**: `icon-8`, `icon-12`, `icon-16`, `icon-20`, `icon-24`, `icon-40`, `icon-full-size` 

**Rotations**:  `rotate-0`, `rotate-45`, `rotate-90`, `rotate-135`, `rotate-180`, `rotate-225`, `rotate-270`, `rotate-315`

<!-- Auto Generated Below -->


## Usage

### Angular

```html
<!-- Default -->
<kv-svg-icon name="kv-logo-kelvin"></kv-svg-icon>

<!-- Custom CSS -->
<kv-svg-icon name="kv-logo-kelvin" customClass="icon-24 rotate-90"><kv-svg-icon>

<!-- Custom Color -->
<kv-svg-icon name="kv-logo-kelvin" customColor="#103d73"><kv-svg-icon>
```


### React

```tsx
import React from 'react';

import { KvSvgIcon } from '@kelvininc/react-ui-components';

export const SvgIconExample: React.FC = () => (
	<>
		{/*-- Default --*/}
		<KvSvgIcon name="kv-logo-kelvin" />

		{/*-- Custom CSS --*/}
		<KvSvgIcon name="kv-logo-kelvin" customClass="icon-24 rotate-90" />

		{/*-- Custom Color --*/}
		<KvSvgIcon name="kv-logo-kelvin" customColor="#103d73" />
	</>
);
```



## Properties

| Property            | Attribute      | Description                                                                                                                 | Type                 | Default     |
| ------------------- | -------------- | --------------------------------------------------------------------------------------------------------------------------- | -------------------- | ----------- |
| `assetsUrl`         | `assets-url`   | (required) Url where assets are public served                                                                               | `string`             | `'/'`       |
| `customClass`       | `custom-class` | (optional) Additional classes to apply for custom CSS. If multiple classes are provided they should be separated by spaces. | `string \| string[]` | `''`        |
| `customColor`       | `custom-color` | (optional) Icon custom color                                                                                                | `string`             | `''`        |
| `name` _(required)_ | `name`         | (required) Icon symbol name                                                                                                 | `string`             | `undefined` |


## Shadow Parts

| Part     | Description         |
| -------- | ------------------- |
| `"icon"` | The icon container. |


## CSS Custom Properties

| Name                            | Description                                        |
| ------------------------------- | -------------------------------------------------- |
| `--icon-color`                  | Color of the icon                                  |
| `--icon-height`                 | Height of icon                                     |
| `--icon-rotation`               | Rotation of icon                                   |
| `--icon-width`                  | Width of icon                                      |
| `--illustration-color`          | Base color used in the illustration                |
| `--illustration-color-dark-35`  | 35% darker version of the illustration base color  |
| `--illustration-color-dark-55`  | 55% darker version of the illustration base color  |
| `--illustration-color-light-80` | 80% lighter version of the illustration base color |
| `--illustration-color-light-90` | 90% lighter version of the illustration base color |
| `--illustration-success-color`  | Success color used in the illustration             |


## Dependencies

### Used by

 - [kv-switch-button](../switch-button)

### Graph
```mermaid
graph TD;
  kv-switch-button --> kv-svg-icon
  style kv-svg-icon fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


