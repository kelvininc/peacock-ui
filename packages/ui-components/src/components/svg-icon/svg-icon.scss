@import '../../assets/styles/globals';

:host {
	/**
	 * @prop --icon-color: Color of the icon
	 * @prop --icon-height: Height of icon
	 * @prop --icon-width: Width of icon
	 * @prop --icon-rotation: Rotation of icon
	 */
	--icon-color: #{kv-color('text')};
	--icon-height: 16px;
	--icon-width: 16px;
	--icon-rotation: 0deg;
}

.svg-icon {
	/**
	 * @prop --illustration-color: Base color used in the illustration
	 * @prop --illustration-color-light-90: 90% lighter version of the illustration base color
	 * @prop --illustration-color-light-80: 80% lighter version of the illustration base color
	 * @prop --illustration-color-dark-55: 55% darker version of the illustration base color
	 * @prop --illustration-color-dark-35: 35% darker version of the illustration base color
	 * @prop --illustration-success-color: Success color used in the illustration
	 */
	--illustration-color: var(--kv-illustrations-color, #{$primary-color});
	--illustration-color-light-90: var(--kv-illustrations-color-light-90, #{mix(#fff, $primary-color, 89.5%)});
	--illustration-color-light-80: var(--kv-illustrations-color-light-80, #{mix(#fff, $primary-color, 78.5%)});
	--illustration-color-dark-55: var(--kv-illustrations-color-dark-55, #{mix(#000, $primary-color, 54.5%)});
	--illustration-color-dark-35: var(--kv-illustrations-color-dark-35, #{mix(#000, $primary-color, 36%)});
	--illustration-success-color: var(--kv-illustrations-success-color, #{$success-color});

	fill: var(--icon-color);
	width: var(--icon-width);
	height: var(--icon-height);
	transform: rotate(var(--icon-rotation));

	/* SIZES */
	&.icon-8 {
		width: 8px;
		height: 8px;
	}

	&.icon-12 {
		width: 12px;
		height: 12px;
	}

	&.icon-16 {
		width: 16px;
		height: 16px;
	}

	&.icon-20 {
		width: 20px;
		height: 20px;
	}

	&.icon-24 {
		width: 24px;
		height: 24px;
	}

	&.icon-40 {
		width: 40px;
		height: 40px;
	}

	&.icon-full-size {
		width: 100%;
		height: 100%;
	}


	/* ROTATIONS */
	&.rotate-0 {
		transform: rotate(0deg);
	}

	&.rotate-45 {
		transform: rotate(45deg);
	}

	&.rotate-90 {
		transform: rotate(90deg);
	}

	&.rotate-135 {
		transform: rotate(135deg);
	}

	&.rotate-180 {
		transform: rotate(180deg);
	}

	&.rotate-225 {
		transform: rotate(225deg);
	}

	&.rotate-270 {
		transform: rotate(270deg);
	}

	&.rotate-315 {
		transform: rotate(315deg);
	}

}
