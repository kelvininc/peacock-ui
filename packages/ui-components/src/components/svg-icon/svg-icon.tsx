import { Component, Prop, Host, h } from '@stencil/core';
import { getClassMap } from '../../utils/css-class.helper';

/**
 * @part icon - The icon container.
 */
@Component({
	tag: 'kv-svg-icon',
	styleUrl: 'svg-icon.scss',
	shadow: true
})
export class KvSvgIcon {
	/** (required) Icon symbol name */
	@Prop({ reflect: true }) name!: string;

	/** (required) Url where assets are public served */
	@Prop({ reflect: true }) assetsUrl: string = '';

	/**
	 * (optional) Additional classes to apply for custom CSS. If multiple classes are
	 * provided they should be separated by spaces.
	 */
	@Prop({ reflect: true }) customClass: string | string[] = '';

	/** (optional) Icon custom color */
	@Prop({ reflect: true }) customColor: string = '';

	render() {
		return (
			<Host>
				<svg
					part="icon"
					xmlns="http://www.w3.org/2000/svg"
					class={{
						...getClassMap(this.customClass),
						'svg-icon': true
					}}
					style={{ fill: this.customColor }}
				>
					<use href={`${this.assetsUrl}svg-symbols.svg#${this.name}`}></use>
				</svg>
			</Host>
		);
	}
}
