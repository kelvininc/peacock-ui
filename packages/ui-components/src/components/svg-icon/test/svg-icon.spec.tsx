import { SpecPage } from '@stencil/core/internal';
import { newSpecPage } from '@stencil/core/testing';
import { KvSvgIcon } from '../svg-icon';

describe('Svg Icon (unit tests)', () => {
	let page: SpecPage;
	let component: KvSvgIcon;

	describe('when uses default props', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvSvgIcon],
				html: `<kv-svg-icon name="kv-logo-kelvin"></kv-svg-icon>`
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});

	describe('when has custom classes', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvSvgIcon],
				html: `<kv-svg-icon name="kv-logo-kelvin" custom-class="icon-full-size rotate-90"></kv-svg-icon>`
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});

	describe('when has custom color', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvSvgIcon],
				html: `<kv-svg-icon name="kv-logo-kelvin" custom-color="#05a357"></kv-svg-icon>`
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});
});
