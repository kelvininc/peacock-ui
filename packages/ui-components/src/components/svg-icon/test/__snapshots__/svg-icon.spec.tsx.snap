// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Svg Icon (unit tests) when has custom classes should match the snapshot 1`] = `
<kv-svg-icon custom-class="icon-full-size rotate-90" custom-color="" name="kv-logo-kelvin">
  <mock:shadow-root>
    <svg class="icon-full-size rotate-90 svg-icon" part="icon" xmlns="http://www.w3.org/2000/svg">
      <use href="/svg-symbols.svg#kv-logo-kelvin"></use>
    </svg>
  </mock:shadow-root>
</kv-svg-icon>
`;

exports[`Svg Icon (unit tests) when has custom color should match the snapshot 1`] = `
<kv-svg-icon custom-class="" custom-color="#05a357" name="kv-logo-kelvin">
  <mock:shadow-root>
    <svg class="svg-icon" part="icon" xmlns="http://www.w3.org/2000/svg" style="fill: #05a357;">
      <use href="/svg-symbols.svg#kv-logo-kelvin"></use>
    </svg>
  </mock:shadow-root>
</kv-svg-icon>
`;

exports[`Svg Icon (unit tests) when uses default props should match the snapshot 1`] = `
<kv-svg-icon custom-class="" custom-color="" name="kv-logo-kelvin">
  <mock:shadow-root>
    <svg class="svg-icon" part="icon" xmlns="http://www.w3.org/2000/svg">
      <use href="/svg-symbols.svg#kv-logo-kelvin"></use>
    </svg>
  </mock:shadow-root>
</kv-svg-icon>
`;
